### PR TITLE
Improve VF2Layout pass

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit fixes some issues with the initial implementation of the
VF2Layout pass and improves the pass to make it suitable for use in the
preset pass managers. This includes several changes:

1. The vf2 pass was raising an exception if it was
   run with >2 qubit gates. This caused issues if we run with
   calibrations (or backends that support >2q gates) as vf2 layout is
   used as an opportunistic thing and if there is for example a 5q gate
   being used we shouldn't fail the entire transpile just because vf2
   can't deal with it. It's only an issue if the later passes can't
   either, it just means vf2 won't be able to find a perfect layout.

2. Add multiple trials to the pass and a mapping quality heuristic with
   basic noise awareness to the pass. For performance we should be using
   the vf2++ heuristic as it improves the performance and makes matches
   much easier to find, however this has the tradeoff as the vf2++
   heuristic orders the nodes by degree for the search which means when
   there are multiple mappings found (as in the case of a 1q gate only
   circuit) it would favor the qubits with highest degree which isn't
   desirable as these tend to be the noisiest qubits. Now the pass will
   iterate over all the mappings found by `retworkx.vf2_mapping()` and
   score each of them, picking the best mapping. If a `BackendProperties`
   object is available it will factor in readout error for each qubit,
   if it's not it scores nodes with lower degree more favorably. A
   future optimization would be to look at 1q and 2q error rates on the
   device and pick based on that, but this gets tricky as to do this
   well we should weight this based on the number of operations on the
   qubits or edges which isn't as simple, so this is left for a future
   optimization

3. Add configurable limits to the pass. While VF2 is normally quite
   quick both to find mappings and to also determine if a mapping isn't
   feasible it isn't always. The edge cases can take quite a long time.
   If we're going to use this pass in the preset pass managers we need
   to be able to set limits on the execution. This commit enables the
   pass to be configured with limits on the number of trials to
   consider, how many internal state visits to use in vf2 before we give
   up (which can be used as a proxy for per trial timeout 1e7 is roughly
   20 seconds), and a total timeout which will be checked after each
   trial to ensure the total time spent in the pass doesn't exceed a
   certain amount of time. This gives us the necessary knobs to ensure
   that the pass doesn't spend too much time trying to find an ideal
   solution.

4. Using an Enum class for the stop conditions. Since there are a finite
   number of stop conditions for the pass, instead of using magic
   strings to determine this using an enum makes it easier for users to
   programmatically interact with the pass and it's return status. The
   value for the enums are the same strings as used before but this
   gives a slightly cleaner interface for interacting with the
   VF2LayoutStopReason value in the property set.

5. Add debug log messages to the pass. After changes 1-4 the internal
    logic and operation of the pass is sufficiently complicated that some
    logging is necessary to figure out how it's behaving. This commit
    adds a bunch of log messages to describe the operation as the pass
    operates all at the debug level because it's all only useful for
    debugging how the pass could have gone wrong and not generally useful
    for information for users.

### Details and comments

This was originally split out from #7213 where these changes were
originally made incrementally as we experimented with making this pass
run by default for optimization levels >= 1.

### CODE REVIEW

**Comment by mtreinish** [comment]
The one thing I'm still debating is removing the node order randomization, mostly because I think it's probably unnecessary as `id_order=False` reorders the nodes by degree and reordering the nodes before we call `vf2_mapping` just eats time.

File: qiskit/transpiler/passes/layout/vf2_layout.py [hunk 1]
```diff
@@ -27,26 +43,50 @@ class VF2Layout(AnalysisPass):
...
 
         Args:
             coupling_map (CouplingMap): Directed graph representing a coupling map.
             strict_direction (bool): If True, considers the direction of the coupling map.
                                      Default is False.
             seed (int): Sets the seed of the PRNG. -1 Means no node shuffling.
+            call_limit (int): The number of state visits to attempt in each execution of
+                vf2.
+            time_limit (float): The total time limit in seconds to run vf2layout
+            properties (BackendProperties): The backend properties for the backend
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/vf2_layout.py:75]:
```suggestion
            properties (BackendProperties): The backend properties for the backend. If ``properties.readout_error()`` is available,  it is used to score the layout.
```

File: qiskit/transpiler/passes/layout/vf2_layout.py [hunk 2]
```diff
@@ -27,26 +43,50 @@ class VF2Layout(AnalysisPass):
...
+    ):
         """Initialize a ``VF2Layout`` pass instance
 
         Args:
             coupling_map (CouplingMap): Directed graph representing a coupling map.
             strict_direction (bool): If True, considers the direction of the coupling map.
                                      Default is False.
             seed (int): Sets the seed of the PRNG. -1 Means no node shuffling.
+            call_limit (int): The number of state visits to attempt in each execution of
+                vf2.
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/vf2_layout.py:73]:
```suggestion
                VF2.
```

File: qiskit/transpiler/passes/layout/vf2_layout.py [hunk 3]
```diff
@@ -27,26 +43,50 @@ class VF2Layout(AnalysisPass):
...
+            time_limit (float): The total time limit in seconds to run vf2layout
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/vf2_layout.py:74]:
```suggestion
            time_limit (float): The total time limit in seconds to run ``VF2Layout``
```

File: qiskit/transpiler/passes/layout/vf2_layout.py [hunk 4]
```diff
@@ -27,26 +43,50 @@ class VF2Layout(AnalysisPass):
...
+            properties (BackendProperties): The backend properties for the backend
+            max_trials (int): The maximum number of trials to run vf2 to find
```
Comment (1ucian0) [inline qiskit/transpiler/passes/layout/vf2_layout.py:76]:
```suggestion
            max_trials (int): The maximum number of trials to run VF2 to find
```

File: qiskit/transpiler/passes/layout/vf2_layout.py (hunk 2 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/vf2_layout.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7276/commits/6b82d087c4c94f294940a31e23053c0cf4294aa8

File: qiskit/transpiler/passes/layout/vf2_layout.py (hunk 3 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/vf2_layout.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7276/commits/6b82d087c4c94f294940a31e23053c0cf4294aa8

File: qiskit/transpiler/passes/layout/vf2_layout.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/vf2_layout.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7276/commits/6b82d087c4c94f294940a31e23053c0cf4294aa8

File: qiskit/transpiler/passes/layout/vf2_layout.py (hunk 4 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/layout/vf2_layout.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7276/commits/6b82d087c4c94f294940a31e23053c0cf4294aa8

**Review by 1ucian0 — APPROVED** [review]
Thanks!

**Pull request merged**